hacks/hexadrop.man
.TH XScreenSaver 1 "" "X Version 11"
.SH NAME
hexadrop \- shrinking hexagons.
.SH SYNOPSIS
.B hexadrop
[\-\-display \fIhost:display.screen\fP]
[\-\-visual \fIvisual\fP]
[\-\-window]
[\-\-root]
[\-\-window\-id \fInumber\fP]
[\-\-delay \fInumber\fP]
[\-\-speed \fInumber\fP]
[\-\-size \fInumber\fP]
[\-\-sides \fInumber\fP]
[\-\-uniform-speed]
[\-\-no-uniform-speed]
[\-\-lockstep]
[\-\-no-lockstep]
[\-\-fps]
.SH DESCRIPTION
Draws a grid of hexagons or other shapes and drops them out.
.SH OPTIONS
.TP 8
.B \-\-visual \fIvisual\fP
Specify which visual to use.  Legal values are the name of a visual class,
or the id number (decimal or hex) of a specific visual.
.TP 8
.B \-\-window
Draw on a newly-created window.  This is the default.
.TP 8
.B \-\-root
Draw on the root window.
.TP 8
.B \-\-window\-id \fInumber\fP
Draw on the specified window.
.TP 8
.B \-\-delay \fInumber\fP
Per-frame delay, in microseconds.  Default: 30000 (0.03 seconds.).
.TP 8
.B \-\-speed \fInumber\fP
Speed. 0.5 for half as fast; 2.0 for twice as fast.
.TP 8
.B \-\-size \fInumber\fP
How many tiles to fit horizontally on the screen.  Default 15.
.TP 8
.B \-\-sides \fInumber\fP
Shape of the tiles. 3, 4, 6 or 8. Default: random.
.TP 8
.B \-\-uniform-speed | \-\-no-uniform-speed
Whether each tile should drop at the same speed. Default: random.
.TP 8
.B \-\-lockstep | \-\-no-lockstep
Whether each tile should drop at the same time. Default: random.
.TP 8
.B \-\-fps | \-\-no-fps
Whether to show a frames-per-second display at the bottom of the screen.
.SH ENVIRONMENT
.PP
.TP 8
.B DISPLAY
to get the default host and display number.
.TP 8
.B XENVIRONMENT
to get the name of a resource file that overrides the global resources
stored in the RESOURCE_MANAGER property.
.TP 8
.B XSCREENSAVER_WINDOW
The window ID to use with \fI\-\-root\fP.
.SH SEE ALSO
.BR X (1),
.BR xscreensaver (1)
.SH COPYRIGHT
Copyright \(co 2013 by Jamie Zawinski.  Permission to use, copy, modify, 
distribute, and sell this software and its documentation for any purpose is 
hereby granted without fee, provided that the above copyright notice appear 
in all copies and that both that copyright notice and this permission notice
appear in supporting documentation.  No representations are made about the 
suitability of this software for any purpose.  It is provided "as is" without
express or implied warranty.
.SH AUTHOR
Jamie Zawinski.


po/ChangeLog
2002-11-25  Kostas Papadimas <pkst@gmx.net>

	* el.po: Added Greek translation

2002-11-22  Dmitry G. Mastrukov  <dmitry@taurussoft.org>

	* be.po: Updated Belarusian translation
	* from Belarusian team <i18n@infonet.by>.

2002-11-17  Dmitry G. Mastrukov  <dmitry@taurussoft.org>

	* be.po: Added Belarusian translation
	* from Belarusian team <i18n@infonet.by>.

2002-11-06  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Updated Spanish translation.

2002-09-24  Christian Meyer  <chrisime@gnome.org>

	* de.po: Updated German translation.

2002-09-23  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Fixed the (C) issue in the english message.

2002-09-22  Pablo Saratxaga  <pablo@mandrakesoft.com>

	* vi.po: Updated Vietnamese file

2002-09-20  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Fixed the (C) issue in the english message.

2002-09-19  Pablo Saratxaga  <pablo@mandrakesoft.com>

	* vi.po: Updated Vietnamese file

2002-09-17  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Fixed the (C) issue in the english message.

2002-09-16  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Updated Spanish translation.

2002-09-08  Vincent van Adrighem <V.vanAdrighem@dirck.mine.nu>

        * nl.po: Dutch translation updated by Tino Meinen.
			This should fix the "error" on the statuspages.

2002-09-07  Christian Neumair  <chris@gnome-de.org>

	* de.po: Updated German translation.

2002-09-05  Christian Neumair  <chris@gnome-de.org>

	* de.po: Updated German translation.

2002-09-03  Kjartan Maraas  <kmaraas@gnome.org>

	* no.po: Updated Norwegian (bokmål) translation.

2002-08-29  Ole Laursen  <olau@hardworking.dk>

	* da.po: Updated Danish translation.

2002-08-29  Tõivo Leedjärv  <toivo@linux.ee>

	* et.po: Fixed Estonian translation.

2002-08-29  Christian Neumair  <christian-neumair@web.de>

	* de.po: Updated German translation.

2002-08-26  Ole Laursen  <olau@hardworking.dk>

	* da.po: Partial update of Danish translation.

2002-08-23  Andras Timar  <timar@gnome.huy

	* hu.po: Added Hungarian translation (partial).

2002-08-13  Christian Neumair  <christian-neumair@web.de>

	* de.po: Started German translation.

2002-08-11  Vincent van Adrighem <V.vanAdrighem@dirck.mine.nu>

        * nl.po: Dutch translation updated by Tino Meinen.

2002-07-28  Christophe Merlet <christophe@merlet.net>

	* fr.po: Updated French translation from contribution of
	Sun G11n <gnome_int_l10n@ireland.sun.com>.

2002-07-22  Christian Rose  <menthos@menthos.com>

	* sv.po: Updated Swedish translation somewhat.

2002-07-12  Pablo Saratxaga  <pablo@mandrakesoft.com>

	* vi.po,wa.po: Updated Vietnamese and Walloon files

2002-06-26  Carlos Perello Marin <carlos@gnome-db.org>

	* es.po: Fixed a duplicated string.

2002-06-24  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Updated Spanish Translation.

2002-06-25  Christian Rose  <menthos@menthos.com>

	* fr.po, pl.po, ru.po, uk.po, wa.po: Added from xscreensaver module.

2002-06-21  Changwoo Ryu  <cwryu@debian.org>

	* ko.po: Added Korean translation from Young-Ho Cha
	<ganadist@chollian.net>.

2002-06-19  Christian Rose  <menthos@menthos.com>

	* sv.po: Updated Swedish translation somewhat.

2002-06-18  Carlos Perello Marin <carlos@gnome-db.org>

	* INFO, xscreensaver.pot: Updated from 4.0.5 release.
	
2002-06-16  Carlos Perello Marin <carlos@gnome-db.org>

	* INFO: Added extra info. I will take all translations from here, you
	don't need to send me it by mail.

2002-06-16  Pablo Saratxaga  <pablo@mandrakesoft.com>

	* vi.po: Added Vietnamese file

2002-06-14  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Updated Spanish Translation.

2002-06-14  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Updated Spanish Translation.

2002-06-12  Abel Cheung  <maddog@linux.org.hk>

	* zh_TW.po: New traditional Chinese translation.

2002-06-09  Takeshi Aihana <aihana@gnome.gr.jp>

	* ja.po: Added as Japanese translation.

2002-06-06  Christian Rose  <menthos@menthos.com>

	* sv.po: Added from xscreensaver module.

2002-06-06  Ole Laursen  <olau@hardworking.dk>

	* da.po: Added and halfway-updated from xscreensaver module.

2002-06-06  TÃµivo LeedjÃ€rv  <toivo@linux.ee>

	* et.po: Added from xscreensaver module.

2002-06-06  Carlos Perello Marin <carlos@gnome-db.org>

	* es.po,nl.po,no.po,pt.po,sk.po,zh_CN.po: Added from xscreensaver
	module.
	* xscreensaver.pot: Added from latest release (4.04.2).
	* INFO: Added info about this module.

2002-06-04  Vincent van Adrighem <V.vanAdrighem@dirck.mine.nu>

        * nl.po: Added Dutch translation.

2002-05-26  Pablo Gonzalo del Campo

	* es.po: Updated Spanish Translation.

2002-05-25  Pablo Gonzalo del Campo

	* es.po: Updated Spanish Translation.

2002-05-23  Pablo Gonzalo del Campo

	* es.po: Updated Spanish Translation.

2002-05-23  Wang Jian  <lark@linux.net.cn>

	* zh_CN.po: Backported Simplified Chinese translation from HEAD.

2002-05-19  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Updated Spanish Translation.
						
2002-05-17  Wang Jian  <lark@linux.net.cn>

	* zh_CN.po: Added Simplified Chinese translation by
	He Qiangqiang <carton@linux.net.cn>.

2002-05-11  Stanislav Visnovsky  <visnovsky@nenya.ms.mff.cuni.cz>

        * sk.po: Backported Slovak translation from HEAD.
	
2002-05-01  GermÃ¡n Poo-CaamaÃ±o  <gpoo@ubiobio.cl>

	* es.po: Updated Spanish translation from
	Guillermo Movia <gmovia@hotpop.com> and converted to UTF-8

2002-04-29  Pablo Saratxaga  <pablo@mandrakesoft.com>

	* wa.po: Updated Walloon file

2002-04-27  Duarte Loreto <happyguy_pt@hotmail.com>

        * pt.po: Added Portuguese translation.

2002-04-26  TÃµivo LeedjÃ€rv  <toivo@linux.ee>

        * et.po: Updated Estonian translation.

2002-04-26  Ole Laursen  <olau@hardworking.dk>

	* da.po: Updated Danish translation.

2002-04-21  Kjartan Maraas  <kmaraas@gnome.org>

	* no.po: Updated Norwegian (bokmÃ¥l) translation.


hacks/hexadrop.c
/* xscreensaver, Copyright (c) 1999-2019 Jamie Zawinski <jwz@jwz.org>
 *
 * Permission to use, copy, modify, distribute, and sell this software and its
 * documentation for any purpose is hereby granted without fee, provided that
 * the above copyright notice appear in all copies and that both that
 * copyright notice and this permission notice appear in supporting
 * documentation.  No representations are made about the suitability of this
 * software for any purpose.  It is provided "as is" without express or 
 * implied warranty.
 *
 * Draws a grid of hexagons or other shapes and drops them out.
 * Created 8-Jul-2013.
 */

#include <math.h>
#include "screenhack.h"

#define countof(x) (sizeof(x)/sizeof(*(x)))
#define ABS(x) ((x)<0?-(x):(x))

/* Avoid rounding errors by using a larger fixed-point grid.
   Without this, we got little pointy errors at some corners. */
#define SCALE 10

typedef struct {
  int sides;
  int cx, cy;
  double th, radius, i, speed;
  int colors[2];
  Bool initted_p;
} cell;

typedef struct {
  Display *dpy;
  Window window;
  XWindowAttributes xgwa;

  int ncells, cells_size, gw, gh;
  cell *cells;

  int delay;
  double speed;
  int sides;
  Bool lockstep_p;
  Bool uniform_p;
  Bool initted_p;

  int ncolors;
  XColor *colors;
  GC gc;

} state;


static void
make_cells (state *st)
{
  int grid_size = get_integer_resource (st->dpy, "size", "Size");
  cell *cells2;
  int size, r, gw, gh, x, y, i;
  double th = 0;

  if (grid_size < 5) grid_size = 5;

  size = ((st->xgwa.width > st->xgwa.height
           ? st->xgwa.width : st->xgwa.height)
          / grid_size);
  gw = st->xgwa.width  / size;
  gh = st->xgwa.height / size;

  switch (st->sides) {
  case 8:
    r  = size * 0.75;
    th = M_PI / st->sides;
    gw *= 1.25;
    gh *= 1.25;
    break;
  case 6:
    r  = size / sqrt(3);
    th = M_PI / st->sides;
    gh *= 1.2;
    break;
  case 3:
    size *= 2;
    r  = size / sqrt(3);
    th = M_PI / st->sides / 2;
    break;
  case 4:
    size /= 2;
    r  = size * sqrt (2);
    th = M_PI / st->sides;
    break;
  default:
    abort();
    break;
  }

  gw += 3;	/* leave a few extra columns off screen just in case */
  gh += 3;

  st->ncells = gw * gh;

  if (st->initted_p && !st->cells) abort();
  if (!st->initted_p && st->cells) abort();

  cells2 = (cell *) calloc (st->ncells, sizeof(*cells2));
  if (! cells2) abort();

  if (st->cells)
    {
      for (y = 0; y < (st->gh < gh ? st->gh : gh); y++)
        for (x = 0; x < (st->gw < gw ? st->gw : gw); x++)
          cells2[y * gw + x] = st->cells [y * st->gw + x];
      free (st->cells);
      st->cells = 0;
    }

  st->cells = cells2;
  st->gw = gw;
  st->gh = gh;

  i = 0;
  for (y = 0; y < gh; y++)
    for (x = 0; x < gw; x++)
      {
        cell *c = &st->cells[i];
        c->sides = st->sides;
        c->radius = SCALE * r;
        c->th = th;

        switch (st->sides) {
        case 8:
          if (x & 1)
            {
              c->cx = SCALE * x * size;
              c->radius /= 2;
              c->th = M_PI / 4;
              c->sides = 4;
              c->radius *= 1.1;
            }
          else
            {
              c->cx = SCALE * x * size;
              c->radius *= 1.02;
              c->radius--;
            }

          if (y & 1)
            c->cx -= SCALE * size;

          c->cy = SCALE * y * size;

         break;
        case 6:
          c->cx = SCALE * x * size;
          c->cy = SCALE * y * size * sqrt(3)/2;
          if (y & 1)
            c->cx -= SCALE * size * 0.5;
          break;
        case 4:
          c->cx = SCALE * x * size * 2;
          c->cy = SCALE * y * size * 2;
          break;
        case 3:
          c->cx = SCALE * x * size * 0.5;
          c->cy = SCALE * y * size * sqrt(3)/2;
          if ((x & 1) ^ (y & 1))
            {
              c->th = th + M_PI;
              c->cy -= SCALE * r * 0.5;
            }
          break;
        default:
          abort();
        }

        if (! c->initted_p)
          {
            c->speed = st->speed * (st->uniform_p ? 1 : (0.1 + frand(0.9)));
            c->i = st->lockstep_p ? 0 : random() % r;
            c->colors[0] = (st->lockstep_p ? 0 : random() % st->ncolors);
            c->colors[1] = 0;
            c->initted_p = True;
          }

        c->radius += SCALE;  /* Avoid single-pixel erase rounding errors */

        if (c->i > c->radius) c->i = c->radius;
        if (c->colors[0] >= st->ncolors) c->colors[0] = st->ncolors-1;
        if (c->colors[1] >= st->ncolors) c->colors[1] = st->ncolors-1;

        i++;
      }

  st->initted_p = True;
}


static void
draw_cell (state *st, cell *c)
{
  XPoint points[20];
  int i, j;
  for (j = 0; j <= 1; j++)
    {
      int r = (j == 0 ? c->radius : c->i);
      for (i = 0; i < c->sides; i++)
        {
          double th = i * M_PI * 2 / c->sides;
          points[i].x = (c->cx + r * cos (th + c->th) + 0.5) / SCALE;
          points[i].y = (c->cy + r * sin (th + c->th) + 0.5) / SCALE;
        }
      XSetForeground (st->dpy, st->gc, st->colors[c->colors[j]].pixel);
      XFillPolygon (st->dpy, st->window, st->gc, points, c->sides,
                    Convex, CoordModeOrigin);
    }

  c->i -= SCALE * c->speed;
  if (c->i < 0)
    {
      c->i = c->radius;
      c->colors[1] = c->colors[0];
      if (c != &st->cells[0])
        c->colors[0] = st->cells[0].colors[0];
      else
        c->colors[0] = random() % st->ncolors;
    }
}


static void
hexadrop_init_1 (Display *dpy, Window window, state *st)
{
  XGCValues gcv;
  char *s1, *s2;

  st->dpy = dpy;
  st->window = window;
  st->delay = get_integer_resource (st->dpy, "delay", "Integer");
  st->ncolors = get_integer_resource (st->dpy, "ncolors", "Integer");
  st->speed = get_float_resource (st->dpy, "speed", "Speed");
  if (st->speed < 0) st->speed = 0;

  XGetWindowAttributes (st->dpy, st->window, &st->xgwa);

  if (st->ncolors < 2) st->ncolors = 2;

  st->colors = (XColor *) calloc (sizeof(*st->colors), st->ncolors);

  if (st->ncolors < 10)
    make_random_colormap (st->xgwa.screen, st->xgwa.visual, st->xgwa.colormap,
                          st->colors, &st->ncolors, False, True, 0, True);
  else
    make_smooth_colormap (st->xgwa.screen, st->xgwa.visual, st->xgwa.colormap,
                          st->colors, &st->ncolors, True, 0, True);
  XSetWindowBackground (dpy, window, st->colors[0].pixel);

  s1 = get_string_resource (st->dpy, "uniform", "Uniform");
  s2 = get_string_resource (st->dpy, "lockstep", "Lockstep");

  if ((!s1 || !*s1 || !strcasecmp(s1, "maybe")) &&
      (!s2 || !*s2 || !strcasecmp(s2, "maybe")))
    {
      /* When being random, don't do both. */
      st->uniform_p = random() & 1;
      st->lockstep_p = st->uniform_p ? 0 : random() & 1;
    }
  else 
    {
      if (!s1 || !*s1 || !strcasecmp(s1, "maybe"))
        st->uniform_p = random() & 1;
      else
        st->uniform_p = get_boolean_resource (st->dpy, "uniform", "Uniform");

      if (!s2 || !*s2 || !strcasecmp(s2, "maybe"))
        st->lockstep_p = random() & 1;
      else
        st->lockstep_p = get_boolean_resource (st->dpy, "lockstep","Lockstep");
    }
  if (s1) free (s1);
  if (s2) free (s2);

  st->sides = get_integer_resource (st->dpy, "sides", "Sides");
  if (! (st->sides == 0 || st->sides == 3 || st->sides == 4 || 
         st->sides == 6 || st->sides == 8))
    {
      printf ("%s: invalid number of sides: %d\n", progname, st->sides);
      st->sides = 0;
    }

  if (! st->sides)
    {
      static int defs[] = { 3, 3, 3,
                            4,
                            6, 6, 6, 6,
                            8, 8, 8 };
      st->sides = defs[random() % countof(defs)];
    }

  make_cells (st);
  gcv.foreground = st->colors[0].pixel;
  st->gc = XCreateGC (dpy, window, GCForeground, &gcv);
}


static void *
hexadrop_init (Display *dpy, Window window)
{
  state *st = (state *) calloc (1, sizeof(*st));
  hexadrop_init_1 (dpy, window, st);
  return st;
}



static unsigned long
hexadrop_draw (Display *dpy, Window window, void *closure)
{
  state *st = (state *) closure;
  int i;

  for (i = 0; i < st->ncells; i++)
    draw_cell (st, &st->cells[i]);

  return st->delay;
}


static void
hexadrop_reshape (Display *dpy, Window window, void *closure, 
                 unsigned int w, unsigned int h)
{
  state *st = (state *) closure;
  XGetWindowAttributes (st->dpy, st->window, &st->xgwa);
  make_cells (st);
}


static void
hexadrop_free_1 (Display *dpy, Window window, void *closure)
{
  state *st = (state *) closure;
  if (st->colors)
    {
      free_colors (st->xgwa.screen, st->xgwa.colormap, st->colors, st->ncolors);
      free (st->colors);
      st->colors = 0;
    }
  if (st->cells) 
    {
      free (st->cells);
      st->cells = 0;
    }
  if (st->gc)
    {
      XFreeGC (st->dpy, st->gc);
      st->gc = 0;
    }

  memset (st, 0, sizeof(*st));
}


static void
hexadrop_free (Display *dpy, Window window, void *closure)
{
  hexadrop_free_1 (dpy, window, closure);
  free (closure);
}


static Bool
hexadrop_event (Display *dpy, Window window, void *closure, XEvent *event)
{
  state *st = (state *) closure;

  if (screenhack_event_helper (dpy, window, event))
    {
      if (random() % 5)		/* Change everything */
        {
          hexadrop_free_1 (st->dpy, st->window, st);
          hexadrop_init_1 (dpy, window, st);
        }
      else			/* Change colors only */
        {
          /* Save the old geometry */
          cell *c = st->cells;
          int n = st->ncells;
          int s = st->sides;
          int i;

          /* Protect it from being freed */
          st->cells = 0;
          hexadrop_free_1 (st->dpy, st->window, st);
          hexadrop_init_1 (dpy, window, st);

          /* Reset the old cells */
          for (i = 0; i < n; i++)
            c[i].initted_p = False;

          /* Re-init, then put them back. */
          free (st->cells);
          st->cells = c;
          st->ncells = n;
          st->sides = s;
        }

      return True;
    }

  return False;
}


static const char *hexadrop_defaults [] = {
  ".background:		black",
  ".foreground:		white",
  "*fpsSolid:		true",
  "*delay:		30000",
  "*sides:		0",
  "*size:		15",
  "*speed:		1.0",
  "*ncolors:		128",
  "*uniform:		Maybe",
  "*lockstep:		Maybe",
#ifdef HAVE_MOBILE
  "*ignoreRotation:     True",
#endif
  0
};

static XrmOptionDescRec hexadrop_options [] = {
  { "-delay",		".delay",	XrmoptionSepArg, 0 },
  { "-sides",		".sides",	XrmoptionSepArg, 0 },
  { "-size",		".size",	XrmoptionSepArg, 0 },
  { "-speed",		".speed",	XrmoptionSepArg, 0 },
  { "-ncolors",		".ncolors",	XrmoptionSepArg, 0 },
  { "-uniform-speed",	".uniform",	XrmoptionNoArg, "True"  },
  { "-no-uniform-speed",".uniform",	XrmoptionNoArg, "False" },
  { "-nonuniform-speed",".uniform",	XrmoptionNoArg, "False" },
  { "-lockstep",	".lockstep",	XrmoptionNoArg, "True"  },
  { "-no-lockstep",	".lockstep",	XrmoptionNoArg, "False" },
  { 0, 0, 0, 0 }
};

XSCREENSAVER_MODULE ("Hexadrop", hexadrop)
